File: src/main/java/com/googongill/aditory/controller/dto/link/request/CreateLinkRequest.java
package com.googongill.aditory.controller.dto.link.request;

import com.googongill.aditory.domain.Category;
import com.googongill.aditory.domain.Link;
import com.googongill.aditory.domain.User;
import com.googongill.aditory.external.chatgpt.dto.AutoCategorizeResult;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import lombok.Builder;
import lombok.Data;

@Data
@Builder
public class CreateLinkRequest {
    @NotNull
    private boolean autoComplete;
    private String title;
    private String summary;
    @NotBlank
    private String url;
    private Long categoryId;

    public Link toEntity(Category category, User user) {
        return new Link(this.title, this.summary, this.url, category, user);
    }

    public Link toEntity(AutoCategorizeResult autoCategorizeResult, Category category, User user) {
        return new Link(autoCategorizeResult.getTitle(), autoCategorizeResult.getSummary(), this.url, category, user);
    }
}

File: src/main/java/com/googongill/aditory/controller/dto/link/request/CreateLinkRequest.kt
package com.googongill.aditory.controller.dto.link.request

import com.googongill.aditory.domain.Category
import com.googongill.aditory.domain.Link
import com.googongill.aditory.domain.User
import com.googongill.aditory.external.chatgpt.dto.AutoCategorizeResult
import jakarta.validation.constraints.NotBlank
import jakarta.validation.constraints.NotNull

data class CreateLinkRequest(
    @NotNull
    val autoComplete: Boolean,
    val title: String? = null,
    val summary: String? = null,
    @NotBlank
    val url:  String,
    val categoryId: Long? = null
) {
    fun toEntity(category: Category?, user: User?): Link {
        return Link(title!!, summary, url, category!!, user!!)
    }

    fun toEntity(autoCategorizeResult: AutoCategorizeResult, category: Category?, user: User?): Link {
        return Link(autoCategorizeResult.title, autoCategorizeResult.summary, url, category!!, user!!)
    }
}
